packages/utils/crates/utils-db/src/rusty.rs

```diff
@@ -0,0 +1,814 @@
+    }
+
+    fn translate_fs_err(prefix: &str, value: JsValue) -> Status {
+        let err_res = serde_wasm_bindgen::from_value(value);
+        if err_res.is_err() {
+            return Status::new(StatusCode::Unknown, &format!("{prefix}: failed to deserialize error info"))
+        }
+
+        let err: FsErr = err_res.unwrap();
+        debug!("io error: {:?}", err);
+
+        if err.code.is_some() {
+            let code = err.code.unwrap();
+            let path = err.path.unwrap_or("n/a".into());
+            match code.as_str() {
+                "ENOENT" => Status::new(StatusCode::NotFound, &format!("{prefix}: path {path} not found")),
+                "EEXIST" => Status::new(StatusCode::AlreadyExists, &format!("{prefix}: path {path} already exists")),
+                _ => Status::new(StatusCode::Unknown, &format!("{prefix}: unknown error {code} at {path}"))
+            }
+        } else {
+            return Status::new(StatusCode::Unknown, &format!("{prefix}: failed without error code"))
+        }
+    }
+
+    impl FileHandle {
+        pub fn open(path: &str, flags: Option<String>) -> rusty_leveldb::Result<Self> {
+            openSync(path, flags.map(JsString::from), None)
+                .map_err(|e| translate_fs_err("could not open file", e))
+                .map(|fd| {
+                    assert!(fd >= 0, "negative fd");
+                    Self(fd)
+                })
+        }
+
+        fn read_from(&self, offset: Option<u32>, dst: &mut [u8]) -> rusty_leveldb::Result<usize> {
+            let mut ubuf = Uint8Array::new_with_length(dst.len() as u32);
+            let read = readSync(self.0, &mut ubuf, 0, dst.len() as u32, offset)
+                .map_err(|e| translate_fs_err("could not read", e))?;
+
+            if read > 0 {
+                ubuf.copy_to(dst);
+                Ok(read as usize)
+            } else if read == 0 {
+                Ok(0)
+            } else {
+                Err(Status::new(StatusCode::IOError, "read error"))
+            }
+        }
+    }
+
+    impl Read for FileHandle {
+        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            self.read_from(None, buf)
+                .map_err(|_| io::ErrorKind::Other.into())
+        }
+    }
+
+    impl Write for FileHandle {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            let arr = Uint8Array::new_with_length(buf.len() as u32);
+            arr.copy_from(buf);
+
+            let written = writeSync(self.0, &arr, 0, Some(buf.len() as u32), None)
+                .map_err(|_| io::Error::from(io::ErrorKind::Other))?;
+
+            if written >= 0 {
+                Ok(written as usize)
+            } else {
+                Err(io::ErrorKind::Other.into())
+            }
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            fsyncSync(self.0).map_err(|_| io::Error::from(io::ErrorKind::Other))
+        }
+    }
+
+    impl RandomAccess for FileHandle {
```

Comment:
This code could actually be part of the PR to the upstream to include a wasm env.

Reply:
Sure, but as a separate one. Not sure if WASM with Node.js is on their roadmap.
The `feature/no-fs` PR however has a bigger chance to get in.